### ISSUE TITLE
More fine-grained version in name?

### ISSUE THREAD

**Issue by jtv** [kiplingw/libpqxx#1]
Given how easy it is to break binary compatibility in a C++ library, is it wise to make the package name include just the "6" from the libpqxx version number?  A simple change like inlining a small function could make a 6.2.6 binary-incompatible with a 6.2.5.

I would suggest including the minor version number as well ("6.2"), or even the full version ("6.2.4") and then making guaranteed-compatible fixes in Debian package updates ("6.2.4-1").

**Comment by kiplingw** [comment]
I think it's not unusual when the API goes through major changes to add the major version number to the package name. This is helpful here because ideally we'd like people to have the 6 series installed in parallel beside the 4 series of packages. Anything that changes within the 6 series in terms of the ABI the SO name will automatically be modified through a collaboration of the Debian tool chain and libtool.

**Comment by jtv** [comment]
That makes sense.  What I'm trying to convey is that it seems perfectly reasonable to me to want to have, say,6.2.x and 6.3.y installed side by side.

**Comment by kiplingw** [comment]
I'll leave it to your discretion because it's your project. My personal view is you want to be able to have multiple versions installed simultaneously only if there are major changes to the API that some upstream package maintainers might have a good reason to use an older version for. If it's as simple as a signature to a few methods changing then it's ok if it breaks because upstream should fix their packages.

I should also let you know that right now the packaging semi-supports multi-arch installation, meaning libpqxx6*:amd64 and libpqxx6*:i386 if requested can be installed simultaneously after you fix a few things. For one, the headers included should be installed to /usr/include/${DEB_HOST_MULTIARCH}/pqxx/. 

The old pqxx-config utility I would suggest getting rid of since it's redundant with pkg-config(1) today which is a build time dependency both in configure.ac and in debian/control.

The pkg-config .pc manifest file needs to be renamed, perhaps to libpqxx6.pc and the name variable inside of it to libpqxx6 so that build systems can distinguish between different major versions of it that can be installed simultaneously. The manifest should also be installed in /usr/lib/x86_64-linux-gnu/pkgconfig/.

There were some issues with the documentation files which I couldn't get to build. I tried configuring with --enable-documentation and running make(1) in doc/, but it doesn't appear to run anything. Whenever that gets figured out, you'll want to make sure those files ship with the package too.

Hope that helps!

**Comment by vorlock** [comment]
`For one, the headers included should be installed to /usr/include/${DEB_HOST_MULTIARCH}/pqxx/.`
As far as I now there is no difference between headers for archs, so IMO above is not needed,

I'm not sure if in this issue are you talking about build config coming from Jeroen with the source source or about what I did for uploading into Debian. Anyway IMO merging those probably would make sense.

**Comment by kiplingw** [comment]
Consider the scenario where a user attempts to install both i386 and amd64 variants of the libpqxx-dev package. A subsequent install of either of the two architecture given one already installed in the system will fail when it attempts to unpack the headers that coincide with the other package's installed location in /usr/include/pqxx.

This is why s. [9.1.1(4)](https://www.debian.org/doc/debian-policy/#file-system-structure) of the DPM encourages installation in /usr/include/<triplet>/pqxx.

**Comment by jtv** [comment]
Ah, the dreaded "file clash" between packages.  Do I need to change anything upstream for this?

Also, I'm intrigued about how pkg-config might replace my pqxx-config script.  Always nice to ditch some code that's no longer needed.  Will pkg-config do the job even if libpqxx is not yet installed?

**Comment by kiplingw** [comment]
Hey Jeroen! You might not have to do anything. The [debhelper](https://wiki.debian.org/Multiarch/Implementation#dh.281.29_and_autotools) scripts should provide values for the [standard directory variables](https://www.gnu.org/software/automake/manual/html_node/Standard-Directory-Variables.html) if I recall.

As for pkg-config, it's really wonderful. You can safely ditch the old pqxx-config script. It's [very easy](https://people.freedesktop.org/~dbn/pkg-config-guide.html) to work with. Normally only the libfoo-dev package is needed to provide the .pc pkg-config manifest. The libfoo-dev package sometimes include a static library for the cases where a developer wants to use that. The libfoo package normally contains the shared library versions which is what 99% of your users only want when they are not developers.

**Comment by jtv** [comment]
Turns out pqxx-config was already marked as deprecated!  There's some documentation to be updated though, particularly for the Windows builds.

**Comment by kiplingw** [comment]
Well there you go! It's not the only package to have some kind ad hoc pkg-config precursor. There's actually quite a few out there, but thankfully none of them necessary anymore. Thank you freedesktop.org.

**Comment by vorlock** [comment]
Just a heads up guys, my fiddling with package broke it, so to sort it out I had to restore versioning I was using before so, @jtv latest upload has binary named `libpqxx-6.2` it provides 6.2.4 and I'm going to stick to the naming convention.

Well at least as long as again I'll think I can make it better with less work required from my self :-)

@kiplingw I'll try to think about making -dev more multiarch, but if you have ideas, pls create MR into [Salsa debian/sid](https://salsa.debian.org/debian/libpqxx/commits/debian/sid)

**Comment by kiplingw** [comment]
Ok cool @vorlock. I probably don't have much more to contribute on that front beyond what the DPM [suggests](https://www.debian.org/doc/manuals/maint-guide/advanced.en.html#multiarch).
